third_party/move/move-vm/runtime/src/storage/ty_layout_converter.rs

```diff
@@ -0,0 +1,506 @@
+                // but their inner types may.
+                let variant_layouts = variants
+                    .iter()
+                    .map(|variant| {
+                        variant
+                            .1
+                            .iter()
+                            .map(|(_, ty)| {
+                                let ty = self
+                                    .vm_config()
+                                    .ty_builder
+                                    .create_ty_with_subst(ty, ty_args)?;
+                                let (ty, has_id_mappings) =
+                                    self.type_to_type_layout_impl(&ty, count, depth)?;
+                                has_identifier_mappings |= has_id_mappings;
+                                Ok(ty)
+                            })
+                            .collect::<PartialVMResult<Vec<_>>>()
+                    })
+                    .collect::<PartialVMResult<Vec<_>>>()?;
+                MoveTypeLayout::Struct(MoveStructLayout::RuntimeVariants(variant_layouts))
+            },
+        };
+
+        let field_node_count = *count - count_before;
+        struct_cache.store_struct_layout_info(
+            struct_name_idx,
+            ty_args.to_vec(),
+            layout.clone(),
+            field_node_count,
+            has_identifier_mappings,
+        );
+        Ok((layout, has_identifier_mappings))
+    }
+
+    fn get_identifier_mapping_kind(
+        &self,
+        idx: StructNameIndex,
+    ) -> PartialVMResult<Option<IdentifierMappingKind>> {
+        if !self.vm_config().delayed_field_optimization_enabled {
+            return Ok(None);
+        }
+        let struct_name = self.struct_name_index_map().idx_to_struct_name(idx)?;
+        Ok(IdentifierMappingKind::from_ident(
+            &struct_name.module,
+            &struct_name.name,
+        ))
+    }
+
+    // -------------------------------------------------------------------------------------
+    // Decorated Layout
+
+    fn type_to_fully_annotated_layout_impl(
+        &self,
+        ty: &Type,
+        count: &mut u64,
+        depth: u64,
+    ) -> PartialVMResult<MoveTypeLayout> {
+        if *count > MAX_TYPE_TO_LAYOUT_NODES {
+            return Err(
+                PartialVMError::new(StatusCode::TOO_MANY_TYPE_NODES).with_message(format!(
+                    "Number of type nodes when constructing type layout exceeded the maximum of {}",
+                    MAX_TYPE_TO_LAYOUT_NODES
+                )),
+            );
+        }
+        if depth > VALUE_DEPTH_MAX {
+            return Err(
+                PartialVMError::new(StatusCode::VM_MAX_VALUE_DEPTH_REACHED).with_message(format!(
+                    "Depth of a layout exceeded the maximum of {} during construction",
+                    VALUE_DEPTH_MAX
+                )),
+            );
+        }
+        Ok(match ty {
+            Type::Bool => MoveTypeLayout::Bool,
```

Comment:
Why we are not incrementing count here (I know you copy-pasted code from before, but why it was not there)?🤦‍♂️ We do it for structs though in `struct_name_to_fully_annotated_layout` ...

Reply:
It's a decision that someone made that primitive types do not count. I agree its weird, but changing this would be a breaking change.